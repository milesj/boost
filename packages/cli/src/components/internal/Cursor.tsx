import React from 'react';
import { Text } from 'ink';
import { Style } from '../Style';

export interface CursorProps {
	focused?: boolean;
	hideCursor?: boolean;
	position: number;
	value: string;
}

export function Cursor({ focused, hideCursor, position, value }: CursorProps) {
	if (!focused) {
		return <Text>{value}</Text>;
	}

	return (
		<Text>
			{value.slice(0, position)}
			{hideCursor ? (
				value[position]
			) : (
				<Style inverted type="info">
					{value[position] || ' '}
				</Style>
			)}
			{value.slice(position + 1)}
		</Text>
	);
}
