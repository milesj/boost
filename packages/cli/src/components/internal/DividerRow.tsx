import React from 'react';
import { Box } from 'ink';
import { Style } from '../Style';

export interface DividerRowProps {
	label: NonNullable<React.ReactNode>;
}

export function DividerRow({ label }: DividerRowProps) {
	return (
		<Box marginLeft={2}>
			<Style type="muted">{label || '────'}</Style>
		</Box>
	);
}
