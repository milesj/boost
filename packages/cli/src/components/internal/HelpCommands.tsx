import { useMemo, useRef } from 'react';
import { Box, Text } from 'ink';
import { stripAnsi } from '@boost/terminal';
import { SPACING_COL, SPACING_COL_WIDE, SPACING_ROW } from '../../constants';
import { formatCommandCall, formatDescription, groupByCategory } from '../../helpers';
import { msg } from '../../translate';
import type { Categories, CommandConfigMap } from '../../types';
import { Header } from '../Header';
import { Style } from '../Style';

export interface HelpCommandsProps {
	categories?: Categories;
	commands: CommandConfigMap;
}

export function HelpCommands({ categories, commands }: HelpCommandsProps) {
	const pathWidth = useRef(0);

	const items = useMemo(
		() =>
			Object.entries(commands).map(([path, command]) => {
				const pathCall = formatCommandCall(path, command);
				const pathCallStripped = stripAnsi(pathCall);

				if (pathCallStripped.length > pathWidth.current) {
					pathWidth.current = pathCallStripped.length;
				}

				return {
					...command,
					name: path,
					path,
					pathCall,
				};
			}),
		[commands],
	);

	const categorizedCommands = useMemo(
		() => groupByCategory(items, categories ?? {}),
		[categories, items],
	);

	const categoryCount = Object.keys(categorizedCommands).length;

	return (
		<>
			<Header label={msg('cli:labelCommands')} />

			{Object.entries(categorizedCommands).map(([key, category], index) => (
				<Box key={key} flexDirection="column">
					{!!category.name && categoryCount > 1 && (
						<Box marginTop={index === 0 ? 0 : SPACING_ROW} paddingLeft={SPACING_COL}>
							<Style bold type="none">
								{category.name}
							</Style>
						</Box>
					)}

					{category.items.map((config) => {
						const desc = formatDescription(config);

						return (
							<Box key={key + config.path} flexDirection="row" paddingLeft={SPACING_COL}>
								<Box flexGrow={0} width={pathWidth.current + SPACING_COL_WIDE}>
									<Text>{config.pathCall}</Text>
								</Box>

								<Box flexGrow={1}>
									<Text wrap="wrap">{desc}</Text>
								</Box>
							</Box>
						);
					})}
				</Box>
			))}
		</>
	);
}
