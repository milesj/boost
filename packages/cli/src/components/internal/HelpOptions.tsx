import { useMemo, useRef } from 'react';
import { Box, Text } from 'ink';
import { OptionConfig, OptionConfigMap } from '@boost/args';
import { stripAnsi } from '@boost/terminal';
import { SPACING_COL, SPACING_COL_WIDE, SPACING_ROW } from '../../constants';
import { formatDescription, formatType, groupByCategory } from '../../helpers';
import { msg } from '../../translate';
import { Categories } from '../../types';
import { Header } from '../Header';
import { Style } from '../Style';

function gatherOptionTags(config: OptionConfig): string[] {
	const tags: string[] = [];

	if (config.count) {
		tags.push(msg('cli:tagCount'));
	}

	if (config.multiple) {
		tags.push(msg('cli:tagMultiple') + (config.arity ? `(${config.arity})` : ''));
	}

	return tags;
}

export interface HelpOptionsProps {
	categories?: Categories;
	options: OptionConfigMap;
}

export function HelpOptions({ categories, options }: HelpOptionsProps) {
	const longWidth = useRef(0);
	const shortWidth = useRef(0);

	const items = useMemo(
		() =>
			Object.entries(options).map(([name, option]) => {
				const shortLabel = option.short ? `-${option.short},` : '';
				let longLabel = `--${option.default === true ? 'no-' : ''}${name}`;

				if (option.type !== 'boolean') {
					longLabel += ' ';
					longLabel += formatType(option);
				}

				const longLabelStripped = stripAnsi(longLabel);
				const shortLabelStripped = stripAnsi(shortLabel);

				if (longLabelStripped.length > longWidth.current) {
					longWidth.current = longLabelStripped.length;
				}

				if (shortLabelStripped.length > shortWidth.current) {
					shortWidth.current = shortLabelStripped.length;
				}

				return {
					...option,
					longLabel,
					name,
					shortLabel,
				};
			}),
		[options],
	);

	const categorizedOptions = useMemo(
		() => groupByCategory(items, categories ?? {}),
		[categories, items],
	);
	const categoryCount = Object.keys(categorizedOptions).length;
	const showShortColumn = shortWidth.current > 0;

	return (
		<>
			<Header label={msg('cli:labelOptions')} />

			{Object.entries(categorizedOptions).map(([key, category], index) => (
				<Box key={key} flexDirection="column">
					{!!category.name && categoryCount > 1 && (
						<Box marginTop={index === 0 ? 0 : SPACING_ROW} paddingLeft={SPACING_COL}>
							<Style bold type="none">
								{category.name}
							</Style>
						</Box>
					)}

					{category.items.map((config) => (
						<Box key={key + config.name} flexDirection="row" paddingLeft={SPACING_COL}>
							{showShortColumn && (
								<Box flexGrow={0} width={shortWidth.current + SPACING_COL}>
									<Text>{config.shortLabel}</Text>
								</Box>
							)}

							<Box flexGrow={0} width={longWidth.current + SPACING_COL_WIDE}>
								<Text>{config.longLabel}</Text>
							</Box>

							<Box flexGrow={1}>
								<Text wrap="wrap">{formatDescription(config, gatherOptionTags(config))}</Text>
							</Box>
						</Box>
					))}
				</Box>
			))}
		</>
	);
}
