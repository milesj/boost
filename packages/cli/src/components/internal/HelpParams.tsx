import React, { useMemo } from 'react';
import { Box, Text } from 'ink';
import { ParamConfig, ParamConfigList } from '@boost/args';
import { SPACING_COL, SPACING_COL_WIDE } from '../../constants';
import { formatDescription, formatType, getLongestWidth } from '../../helpers';
import { msg } from '../../translate';
import { CommandConfig } from '../../types';
import { Header } from '../Header';

function gatherParamTags(config: ParamConfig): string[] {
	const tags: string[] = [];

	if (config.required) {
		tags.push(msg('cli:tagRequired'));
	}

	return tags;
}

export interface HelpParamsProps {
	config?: CommandConfig;
	params: ParamConfigList;
}

export function HelpParams({ config, params }: HelpParamsProps) {
	const labels = useMemo(
		() => params.map((p, index) => `${p.label ?? index} ${formatType(p)}`),
		[params],
	);
	const labelWidth = getLongestWidth(labels);
	const allowVariadic = config?.allowVariadicParams;

	return (
		<>
			<Header label={msg('cli:labelParams')} />

			{params.map((param, index) => {
				if (param.hidden) {
					return null;
				}

				const desc = param ? formatDescription(param, gatherParamTags(param)) : '';

				return (
					<Box key={`${labels[index]}-${index}`} flexDirection="row" paddingLeft={SPACING_COL}>
						<Box flexGrow={0} width={labelWidth + SPACING_COL_WIDE}>
							<Text>{labels[index]}</Text>
						</Box>

						<Box flexGrow={1}>
							<Text wrap="wrap">{desc}</Text>
						</Box>
					</Box>
				);
			})}

			{allowVariadic && (
				<Box paddingLeft={SPACING_COL}>
					<Text>
						{`â€¦${typeof allowVariadic === 'string' ? allowVariadic : ''}`}{' '}
						{formatType({
							multiple: true,
							type: 'string',
						})}
					</Text>
				</Box>
			)}
		</>
	);
}
