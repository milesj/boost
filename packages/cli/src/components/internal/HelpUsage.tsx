import React from 'react';
import { Box, Text } from 'ink';
import { toArray } from '@boost/common';
import { DELIMITER, SPACING_COL } from '../../constants';
import { msg } from '../../translate';
import { Header } from '../Header';

export interface HelpUsageProps {
	delimiter?: string;
	usage?: string[] | string;
}

export function HelpUsage({ delimiter = DELIMITER, usage = '' }: HelpUsageProps) {
	return (
		<>
			<Header label={msg('cli:labelUsage')} />

			{toArray(usage).map((example) => (
				<Box key={example} paddingLeft={SPACING_COL}>
					<Text>{example.startsWith(delimiter) ? example : delimiter + example}</Text>
				</Box>
			))}
		</>
	);
}
