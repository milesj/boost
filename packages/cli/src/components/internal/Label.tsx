import React from 'react';
import { Text } from 'ink';

export interface LabelProps {
	children: NonNullable<React.ReactNode>;
}

export function Label({ children }: LabelProps) {
	if (typeof children === 'string') {
		return <Text bold>{children}</Text>;
	}

	// eslint-disable-next-line react/jsx-no-useless-fragment
	return <>{children}</>;
}
