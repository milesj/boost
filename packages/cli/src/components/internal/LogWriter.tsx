import { useEffect } from 'react';
import { useStdout, useStderr } from 'ink';
import LogBuffer from '../../LogBuffer';

export interface LogWriterProps {
  errBuffer: LogBuffer;
  outBuffer: LogBuffer;
}

export function LogWriter({ errBuffer, outBuffer }: LogWriterProps) {
  const { write: writeErr } = useStderr();
  const { write: writeOut } = useStdout();

  useEffect(() => errBuffer.on(writeErr), [errBuffer, writeErr]);

  useEffect(() => outBuffer.on(writeOut), [outBuffer, writeOut]);

  return null;
}
