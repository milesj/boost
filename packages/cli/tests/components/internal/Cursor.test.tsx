import React from 'react';
import { render } from 'ink-testing-library';
import { Cursor } from '../../../src/components/internal/Cursor';

describe('Cursor', () => {
	it('renders value with no cursor if not focused', () => {
		const { lastFrame } = render(<Cursor position={0} value="Some string of text" />);

		expect(lastFrame()).toMatchSnapshot();
	});

	it('renders no cursor if `hideCursor` is passed', () => {
		const { lastFrame } = render(<Cursor hideCursor position={0} value="Some string of text" />);

		expect(lastFrame()).toMatchSnapshot();
	});

	it('renders cursor at the start', () => {
		const { lastFrame } = render(<Cursor focused position={0} value="Some string of text" />);

		expect(lastFrame()).toMatchSnapshot();
	});

	it('renders cursor at the end', () => {
		const { lastFrame } = render(<Cursor focused position={18} value="Some string of text" />);

		expect(lastFrame()).toMatchSnapshot();
	});

	it('renders cursor in the middle', () => {
		const { lastFrame } = render(<Cursor focused position={8} value="Some string of text" />);

		expect(lastFrame()).toMatchSnapshot();
	});

	it('renders cursor after value', () => {
		const { lastFrame } = render(<Cursor focused position={19} value="Some string of text" />);

		expect(lastFrame()).toMatchSnapshot();
	});
});
