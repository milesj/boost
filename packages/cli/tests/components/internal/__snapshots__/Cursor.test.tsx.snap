// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cursor renders cursor after value 1`] = `"Some string of text[46m[30m [39m[49m"`;

exports[`Cursor renders cursor at the end 1`] = `"Some string of tex[46m[30mt[39m[49m"`;

exports[`Cursor renders cursor at the start 1`] = `"[46m[30mS[39m[49mome string of text"`;

exports[`Cursor renders cursor in the middle 1`] = `"Some str[46m[30mi[39m[49mng of text"`;

exports[`Cursor renders no cursor if \`hideCursor\` is passed 1`] = `"Some string of text"`;

exports[`Cursor renders value with no cursor if not focused 1`] = `"Some string of text"`;
